### DISCUSSION TITLE
Max size of payload when calling dbus method?

### DISCUSSION THREAD

**Discussion by colazzo** [python-sdbus/python-sdbus#58] [General]
Hello,

What would be the max size of payload when calling dbus method?
I encounter an issue when trying to send big amount of data from the client.
It seems to get stuck from 450K bytes and above, on the  moment the client does the call with a string value of that size it get stuck.

When calling the method with same amount of data from D-Spy it seems to work on the same service (service is also using python-dbus). I have actually a use case where I need to be be able to send around 900K bytes of data in one method call.

Do you know of any that kind of issue on client side? Is it limited through a parameters somewhere?

**Comment by igo95862** [comment]
Hello @colazzo,

I don't know. It might be a libsystemd limitation? It might be a python-sdbus limitation? Might be a bus implementation limit?

What was the method type signature you have used? A string `s` or a bytes array `ay`? I will try to give it a test.

Also what version of python-sdbus are you using and how did you install it?

**Reply by igo95862** [reply]
[The D-Bus specification states](https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-messages) 128 MiB is the maximum size of the message. Beyond that you probably want to pass a file descriptor to a pipe. (which would probably be faster for transfer)

**Comment by colazzo** [comment]
It is string type. The thing is that the same full string copied in D-Spy client was working fine.

I am using latest official version 0.11.1.

**Reply by igo95862** [reply]
Did you use  the PyPI package that uses limited API or did you build it from source which uses the unlimited version specific ABI.

They differ in how the strings are serialized.

**Reply by colazzo** [reply]
I am using poetry. 
So I suppose this is PyPi version?
How is this being limited and how to solve?

**Reply by igo95862** [reply]
https://docs.python.org/3.11/c-api/stable.html#limited-c-api

It is limited to allow for forward compatibility. There is only one version uploaded to PyPI built for Python 3.8 but it is compatible with all versions going forward. Unfortunately the Python 3.8 limited api lacks the function calls to convert Python strings to `const char *` so it has to do extra calls.

When installing from PyPI instead of installing the binary package you can download the source package and compile it locally. Unfortunately this means you need to meet all compile time dependencies on your local system. (see README section) I am not familiar with poetry but for pip you can pass `--no-binary` flag to force compile packages.

**Reply by colazzo** [reply]
There seems to be something similar has been added in Poetry as well. I will give it a try. Will give you an update later about it. 

Thanks

**Reply by igo95862** [reply]
Ok. I managed to replicate this. It can be replicated on both limited and unlimited APIs.

It seems like appending to a message is fine but something happens when trying to send it.

**Comment by igo95862** [comment]
Hmmm. I tested the 442_368 length string and it seems to be sometimes working meaning there is probably some kind of race condition?

**Reply by colazzo** [reply]
Don't know exactly on what size length it stopped, but it was around 450K indeed.
Can you debug inside sd_bus_internals_message.c ?

**Reply by igo95862** [reply]
I am doing it right now but it seems like sd-bus never send the message. It does not show up even on bus monitor.

**Reply by igo95862** [reply]
Hmmmmmm.... (strace dump)

```
sendmsg(6, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="l\1\0\1\5\300\6\0\3\0\0\0g\0\0\0\1\1o\0\1\0\0\0/\0\0\0\0\0\0\0"..., iov_len=120}, {iov_base="\0\300\6\0aaaaaaaaaaaaaaaaaaaaaaaaaaaa"..., iov_len=442373}], msg_iovlen=2, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 438528
epoll_wait(3, 0x7f5ba85f28f0, 2, -1)    = -1 EINTR (Interrupted system call)
```

**Reply by igo95862** [reply]
Feels like an sd-bus bug. I will try to make a simple C program tomorrow to replicate.

**Reply by colazzo** [reply]
Ok thanks. Hope we can find some fix or workaround.

**Comment by igo95862** [comment]
@colazzo I think I narrowed down the issue. It has to do something with the epoll and how python-sdbus watches the file descriptor.

I added this code to manually process sd-bus and sending the extremely large string worked:

```diff
diff --git a/src/sdbus/dbus_proxy_async_method.py b/src/sdbus/dbus_proxy_async_method.py
index f7d39c1..a2a0bfa 100644
--- a/src/sdbus/dbus_proxy_async_method.py
+++ b/src/sdbus/dbus_proxy_async_method.py
@@ -85,8 +85,16 @@ class DbusMethodAsyncProxyBind(DbusMethodAsyncBaseBind):
         self.__doc__ = dbus_method.__doc__
 
     async def _dbus_async_call(self, call_message: SdBusMessage) -> Any:
+        from asyncio import sleep
         bus = self.proxy_meta.attached_bus
-        reply_message = await bus.call_async(call_message)
+        fut = bus.call_async(call_message)
+        while True:
+            bus.drive()
+            await sleep(0.1)
+
+            if fut.done():
+                break
+        reply_message = await fut
         return reply_message.get_contents()
 
     @staticmethod
```

**Reply by colazzo** [reply]
Wow, good you found the root cause! 
Just to understand at bit more the issue, was the file handle being closed too early during processing and therefore caused some raise condition? 

Could you put this fix on a separate branch so I can checkout that branch and test it locally?

**Reply by igo95862** [reply]
To enable asynchronous IO sd-bus has a call `sd_bus_get_fd` which returns a file descriptor that the asyncio loop will watch and call the `sd_bus_process` when it is ready. It looks like it has to be monitored not only for a read events but write events as well. I only had it watched for read events.

I am looking in to the best way to integrate it with asyncio. Simply adding a writer seems to raise a few errors as well as  call `sd_bus_process` too many times.

**Comment by igo95862** [comment]
@colazzo I added the 03593a13167a5a0bd6818489d63f99777ab76299 which should fix the issue. I added a unit test to verify the behavior.

You can test it by building it from source `python setup.py build` and then using the `PYTHONPATH` environment variable with the path to where setup tools will place the build. (on my machine the build path is `./build/lib.linux-x86_64-cpython-311/ `)

**Reply by colazzo** [reply]
Hi @igo95862, your fix seems to be working fine, did some test with 1.8 M and data arrived well on service side. 
Well done!
